File: wifi/CommClient-Desktop/src/renderer/services/network/EventEnvelope.ts
/**
 * EventEnvelope — TypeScript mirror of the server-side Pydantic
 * Envelope schema (app/services/event_envelope.py).
 *
 * Why
 * ---
 * The server now wraps every cross-server event in a uniform envelope
 * with traceId, idempotencyKey, hopIndex, etc. The client doesn't
 * (yet) construct envelopes itself — we pass through whatever shape
 * Socket.IO delivers. But we DO want to:
 *
 *   1. Read trace_id off incoming events when present so the dev
 *      console can correlate UI actions with server-side route
 *      traces (TraceReporter).
 *   2. Generate idempotencyKeys for client-initiated commands so
 *      retries on flaky networks don't double-execute.
 *   3. Be ready when the client opts into producing envelopes
 *      directly (Phase 2 — once the server's emit handlers all
 *      accept envelope-shaped payloads as input too).
 *
 * Hard guards mirrored from the server schema
 * -------------------------------------------
 *   * payload size <= 8 KB (control plane only)
 *   * plane = "data" forbidden
 *   * P0 requires_ack = true
 */

export type Priority = 'P0' | 'P1' | 'P2' | 'P3' | 'P4';

export interface Envelope {
  // ── Identity ────────────────────────────────────────────────
  event_id: string;
  trace_id: string;
  span_id: string;
  parent_span_id: string | null;
  idempotency_key: string;

  // ── Type & Priority ────────────────────────────────────────
  event_type: string;
  command_type: string | null;
  priority: Priority;
  plane: 'control';

  // ── Routing ────────────────────────────────────────────────
  source_user_id: string | null;
  destination_user_id: string | null;
  source_server_id: string;
  destination_server_id: string | null;
  current_server_id: string;
  next_server_id: string | null;
  route_id: string | null;
  route_version: number;
  hop_index: number;
  max_hops: number;

  // ── Domain context ─────────────────────────────────────────
  call_id: string | null;
  channel_id: string | null;

  // ── Reliability ────────────────────────────────────────────
  ttl_ms: number;
  sequence: number | null;
  requires_ack: boolean;
  retry_count: number;
  max_retries: number;

  // ── Lifecycle ──────────────────────────────────────────────
  created_at: string;   // ISO-8601
  expires_at: string;   // ISO-8601

  // ── Payload ────────────────────────────────────────────────
  payload: Record<string, unknown>;
}

const MAX_PAYLOAD_BYTES = 8 * 1024;
const MAX_HOPS_DEFAULT = 8;

const PRIORITY_DEFAULT_TTL_MS: Record<Priority, number> = {
  P0: 5_000,
  P1: 30_000,
  P2: 60_000,
  P3: 2_000,
  P4: 10_000,
};

/**
 * Generate a ULID-style ID. Sortable by creation time, 26 chars
 * after the prefix. Uses crypto.getRandomValues so it's safe to
 * collide with server-generated IDs (different prefix space).
 */
function genId(prefix: string): string {
  const tsMs = Date.now() & ((1 << 30) - 1); // 30 bits is plenty
  const tsB32 = tsMs.toString(32).padStart(7, '0');
  const randBytes = new Uint8Array(10);
  crypto.getRandomValues(randBytes);
  let randB32 = '';
  for (let i = 0; i < randBytes.length; i++) {
    randB32 += randBytes[i].toString(32).padStart(2, '0');
  }
  return `${prefix}_${tsB32}${randB32}`.slice(0, prefix.length + 1 + 26);
}

export class PayloadTooLargeError extends Error {
  constructor(size: number) {
    super(
      `payload size ${size} exceeds ${MAX_PAYLOAD_BYTES} bytes — ` +
        `large payloads must travel via S3 (files) or SFU (media)`,
    );
    this.name = 'PayloadTooLargeError';
  }
}

export interface NewEnvelopeArgs {
  event_type: string;
  priority: Priority;
  source_server_id: string;
  payload?: Record<string, unknown>;
  source_user_id?: string;
  destination_user_id?: string;
  destination_server_id?: string;
  call_id?: string;
  channel_id?: string;
  idempotency_key?: string;
  ttl_ms?: number;
  max_hops?: number;
  sequence?: number;
  requires_ack?: boolean;
  max_retries?: number;
  command_type?: string;
  trace_id?: string;
  parent_span_id?: string;
}

/**
 * Construct a fresh envelope. Mirrors `Envelope.new()` on the server.
 * Throws `PayloadTooLargeError` if the resulting JSON exceeds 8 KB.
 */
export function newEnvelope(args: NewEnvelopeArgs): Envelope {
  const now = new Date();
  const ttl = args.ttl_ms ?? PRIORITY_DEFAULT_TTL_MS[args.priority] ?? 5_000;
  const requiresAck = args.priority === 'P0' ? true : args.requires_ack ?? false;

  const env: Envelope = {
    event_id: genId('evt'),
    trace_id: args.trace_id ?? genId('trace'),
    span_id: genId('span'),
    parent_span_id: args.parent_span_id ?? null,
    idempotency_key: args.idempotency_key ?? genId('idem'),
    event_type: args.event_type,
    command_type: args.command_type ?? null,
    priority: args.priority,
    plane: 'control',
    source_user_id: args.source_user_id ?? null,
    destination_user_id: args.destination_user_id ?? null,
    source_server_id: args.source_server_id,
    destination_server_id: args.destination_server_id ?? null,
    current_server_id: args.source_server_id,
    next_server_id: null,
    route_id: null,
    route_version: 1,
    hop_index: 0,
    max_hops: args.max_hops ?? MAX_HOPS_DEFAULT,
    call_id: args.call_id ?? null,
    channel_id: args.channel_id ?? null,
    ttl_ms: ttl,
    sequence: args.sequence ?? null,
    requires_ack: requiresAck,
    retry_count: 0,
    max_retries: args.max_retries ?? 3,
    created_at: now.toISOString(),
    expires_at: new Date(now.getTime() + ttl).toISOString(),
    payload: args.payload ?? {},
  };

  const size = JSON.stringify(env).length;
  if (size > MAX_PAYLOAD_BYTES) {
    throw new PayloadTooLargeError(size);
  }
  return env;
}

export function isExpired(env: Envelope, now: Date = new Date()): boolean {
  return now.getTime() >= new Date(env.expires_at).getTime();
}

/**
 * Look at any incoming socket payload and return the trace_id /
 * event_id if the server tagged it. Older non-fabric events have
 * neither field — caller treats null as "untraced legacy event".
 */
export function readTraceMeta(
  data: unknown,
): { trace_id: string; event_id: string } | null {
  if (data === null || typeof data !== 'object') return null;
  const obj = data as Record<string, unknown>;
  const tid = obj.trace_id;
  const eid = obj.event_id;
  if (typeof tid === 'string' && typeof eid === 'string') {
    return { trace_id: tid, event_id: eid };
  }
  return null;
}

/**
 * Construct an idempotency key for a client-initiated command. Used
 * by handlers that want safe retry semantics on flaky networks.
 *
 * Format: `{action}:{stable_input_hash}` — stable across retries of
 * the same logical action so the server's idempotency cache returns
 * the cached result instead of double-executing.
 */
export function clientIdempotencyKey(action: string, stableInput: string): string {
  return `${action}:${stableInput}`;
}

File: wifi/CommClient-Server/app/services/event_envelope.py
"""
Distributed event envelope — single uniform shape for every event
that traverses server boundaries (broker, federation forwards,
chaos-mode chain, DLQ records).

Why
---
Today every cross-server event has a hand-rolled payload shape and
no consistent way to:

* trace an event end-to-end (no traceId)
* detect duplicates safely (no idempotencyKey on most events)
* age out stale events (no expiresAt)
* prevent loops in chain mode (no hopIndex / maxHops)
* prioritize critical signaling over typing/presence flood (no class)

This module ratifies a single Pydantic model that every server-to-
server event passes through. Backward-compatibility shims convert
legacy dicts at the boundary so old handlers keep working.

Hard guards
-----------
1. ``payload`` size_bytes > 8KB → raise ``PayloadTooLarge``.
   Forces media/files to use S3/SFU instead of the control-plane.
2. ``plane = "data"`` is rejected — control plane only.
3. ``priority = "P0"`` requires ``requires_ack = True``.
4. ``max_hops`` capped at 128 (chaos mode); production default is 8.
5. ``expires_at`` must be after ``created_at`` and roughly equal to
   ``created_at + ttl_ms`` (allow 1s skew for clock drift).
"""

from __future__ import annotations

import base64
import os
import secrets
import time
from datetime import datetime, timedelta, timezone
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field, field_validator, model_validator

# ── Type aliases ────────────────────────────────────────────────

Priority = Literal["P0", "P1", "P2", "P3", "P4"]
Plane = Literal["control", "data"]

# ── Constants ───────────────────────────────────────────────────

MAX_PAYLOAD_BYTES = 8 * 1024  # 8 KB hard cap on control-plane events
# MAX_HOPS_PRODUCTION raised from 24 → 64. With gossip K=10 (default) we
# can reach 10^4 = 10,000 nodes in 4 hops; 64 hops gives massive
# headroom for any realistic mesh, including chains across multiple
# multi-homed bridges. Loop detection (Envelope.is_loop) and the
# trace-id seen-cache still prevent runaway fan-out.
import os as _os_hop
MAX_HOPS_PRODUCTION = max(8, int(_os_hop.environ.get("HELEN_MAX_HOPS", "64")))
MAX_HOPS_CHAOS = 256
DEFAULT_TTL_MS = 5_000

# Default TTLs by priority — callers can override.
PRIORITY_DEFAULT_TTL_MS = {
    "P0": 5_000,    # call signaling — short, must be fresh
    "P1": 30_000,   # call lifecycle
    "P2": 60_000,   # chat messages
    "P3": 2_000,    # presence / typing
    "P4": 10_000,   # file metadata
}

# ── Errors ──────────────────────────────────────────────────────

class EnvelopeError(Exception):
    """Base class for envelope-related errors."""


class PayloadTooLarge(EnvelopeError):
    """Raised when an envelope's payload exceeds MAX_PAYLOAD_BYTES.
    The control-plane explicitly forbids large payloads — they
    should travel out-of-band via SFU or object storage."""


class MaxHopsExceeded(EnvelopeError):
    """Raised when ``Envelope.step()`` would push hop_index beyond
    max_hops. The caller should DLQ the event."""


class LoopDetected(EnvelopeError):
    """Raised when an envelope returns to a server it already passed
    through. Indicates a routing bug or a malicious chain."""


# ── ULID-style ID generator ──
# Avoids pulling in the ulid-py dep just for this. Format is:
#   {prefix}_{ts48}{rand80}     (base32, lowercase)
# Sortable by creation time — same property as ULID. The first 48 bits
# are unix milliseconds; the next 80 bits are crypto-random.

_BASE32_ALPHA = "0123456789abcdefghjkmnpqrstvwxyz"

def _b32(n: int, width: int) -> str:
    out = []
    for _ in range(width):
        out.append(_BASE32_ALPHA[n & 0x1F])
        n >>= 5
    return "".join(reversed(out))

def _gen_id(prefix: str) -> str:
    ts_ms = int(time.time() * 1000) & ((1 << 48) - 1)
    rand = int.from_bytes(os.urandom(10), "big")  # 80 bits
    return f"{prefix}_{_b32(ts_ms, 10)}{_b32(rand, 16)}"


# ── Envelope ────────────────────────────────────────────────────

class Envelope(BaseModel):
    """Uniform envelope wrapping every server-to-server event.

    Construction
    ------------
    Use ``Envelope.new(...)`` for the common case — it generates IDs,
    computes ``expires_at``, and applies priority defaults. Direct
    construction (``Envelope(...)``) is fine but requires every field
    to be supplied. Use ``Envelope.model_validate(dict)`` to parse
    a wire-format envelope received from another server.
    """

    # ── Identity ────────────────────────────────────────────────
    event_id: str
    trace_id: str
    span_id: str
    parent_span_id: Optional[str] = None
    idempotency_key: str

    # ── Type & Priority ────────────────────────────────────────
    event_type: str
    command_type: Optional[str] = None
    priority: Priority
    plane: Plane = "control"

    # ── Routing ────────────────────────────────────────────────
    source_user_id: Optional[str] = None
    destination_user_id: Optional[str] = None
    source_server_id: str
    destination_server_id: Optional[str] = None
    current_server_id: str
    next_server_id: Optional[str] = None
    route_id: Optional[str] = None
    route_version: int = 1
    hop_index: int = 0
    max_hops: int = MAX_HOPS_PRODUCTION

    # ── Domain context ─────────────────────────────────────────
    call_id: Optional[str] = None
    channel_id: Optional[str] = None

    # ── Reliability ────────────────────────────────────────────
    ttl_ms: int = DEFAULT_TTL_MS
    sequence: Optional[int] = None
    requires_ack: bool = False
    retry_count: int = 0
    max_retries: int = 3

    # ── Lifecycle ──────────────────────────────────────────────
    created_at: datetime
    expires_at: datetime

    # ── Payload ────────────────────────────────────────────────
    payload: dict[str, Any] = Field(default_factory=dict)

    # ── Validators ─────────────────────────────────────────────

    @field_validator("event_type")
    @classmethod
    def _valid_event_type(cls, v: str) -> str:
        # Lowercase dotted segments only: "call.signal.offer".
        if not v or "." not in v:
            raise ValueError("event_type must be dotted: e.g. call.signal.offer")
        for seg in v.split("."):
            if not seg or not all(c.islower() or c == "_" or c.isdigit() for c in seg):
                raise ValueError(f"event_type segment {seg!r} must be lowercase a-z/0-9/_")
        return v

    @field_validator("plane")
    @classmethod
    def _no_data_plane(cls, v: Plane) -> Plane:
        if v == "data":
            raise ValueError(
                "plane='data' is forbidden — large/binary payloads must "
                "travel via S3 or SFU, not via envelope."
            )
        return v

    @field_validator("max_hops")
    @classmethod
    def _max_hops_capped(cls, v: int) -> int:
        if v < 1:
            raise ValueError("max_hops must be >= 1")
        if v > MAX_HOPS_CHAOS:
            raise ValueError(
                f"max_hops capped at {MAX_HOPS_CHAOS} (chaos mode). "
                f"Got {v}."
            )
        return v

    @model_validator(mode="after")
    def _consistency(self) -> "Envelope":
        # P0 requires ACK by default — but high-volume signaling
        # (ICE candidates, presence flips) explicitly opts out via
        # ``allow_p0_no_ack``. Without this escape hatch every ICE
        # candidate generates an ACK envelope and the broker doubles
        # in load. ICE candidates are fire-and-forget by WebRTC
        # design (a missed candidate is replaced by gathering on the
        # next pulse), so the ACK is genuinely wasteful.
        if (
            self.priority == "P0"
            and not self.requires_ack
            and not self.payload.get("__allow_p0_no_ack__")
        ):
            raise ValueError(
                "priority=P0 requires requires_ack=True "
                "(or set payload['__allow_p0_no_ack__']=True for ICE-style "
                "fire-and-forget signaling)"
            )

        # expires_at sanity
        if self.expires_at <= self.created_at:
            raise ValueError("expires_at must be after created_at")

        # ttl_ms ↔ expires_at - created_at consistency (allow 1s drift)
        derived_ttl_ms = int(
            (self.expires_at - self.created_at).total_seconds() * 1000
        )
        if abs(derived_ttl_ms - self.ttl_ms) > 1000:
            raise ValueError(
                f"ttl_ms ({self.ttl_ms}) inconsistent with "
                f"expires_at - created_at ({derived_ttl_ms}ms)"
            )

        # hop_index sanity
        if self.hop_index < 0:
            raise ValueError("hop_index must be >= 0")
        if self.hop_index > self.max_hops:
            raise ValueError(
                f"hop_index ({self.hop_index}) exceeds max_hops "
                f"({self.max_hops})"
            )

        # Payload size guard
        size = self.size_bytes()
        if size > MAX_PAYLOAD_BYTES:
            raise PayloadTooLarge(
                f"payload size {size} exceeds {MAX_PAYLOAD_BYTES} bytes — "
                f"large payloads must use S3 (files) or SFU (media)"
            )
        return self

    # ── Construction helpers ───────────────────────────────────

    @classmethod
    def new(
        cls,
        *,
        event_type: str,
        priority: Priority,
        source_server_id: str,
        idempotency_key: Optional[str] = None,
        ttl_ms: Optional[int] = None,
        max_hops: int = MAX_HOPS_PRODUCTION,
        payload: Optional[dict[str, Any]] = None,
        source_user_id: Optional[str] = None,
        destination_user_id: Optional[str] = None,
        destination_server_id: Optional[str] = None,
        call_id: Optional[str] = None,
        channel_id: Optional[str] = None,
        sequence: Optional[int] = None,
        requires_ack: Optional[bool] = None,
        max_retries: int = 3,
        command_type: Optional[str] = None,
        trace_id: Optional[str] = None,
        parent_span_id: Optional[str] = None,
    ) -> "Envelope":
        """Convenience constructor that fills derived fields. Use this
        unless you have a reason to construct manually (e.g. parsing
        wire format)."""
        now = datetime.now(timezone.utc)
        ttl = ttl_ms if ttl_ms is not None else PRIORITY_DEFAULT_TTL_MS.get(priority, DEFAULT_TTL_MS)
        # P0 defaults to requires_ack=True. Caller can explicitly pass
        # ``requires_ack=False`` for fire-and-forget signaling (ICE
        # candidates etc.) — we then add the escape-hatch flag to
        # the payload so schema validation accepts it.
        if priority == "P0":
            if requires_ack is False:
                ack = False
                payload = dict(payload or {})
                payload["__allow_p0_no_ack__"] = True
            else:
                ack = True
        else:
            ack = requires_ack if requires_ack is not None else False

        return cls(
            event_id=_gen_id("evt"),
            trace_id=trace_id or _gen_id("trace"),
            span_id=_gen_id("span"),
            parent_span_id=parent_span_id,
            idempotency_key=idempotency_key or _gen_id("idem"),
            event_type=event_type,
            command_type=command_type,
            priority=priority,
            plane="control",
            source_user_id=source_user_id,
            destination_user_id=destination_user_id,
            source_server_id=source_server_id,
            destination_server_id=destination_server_id,
            current_server_id=source_server_id,
            next_server_id=None,
            route_id=None,
            route_version=1,
            hop_index=0,
            max_hops=max_hops,
            call_id=call_id,
            channel_id=channel_id,
            ttl_ms=ttl,
            sequence=sequence,
            requires_ack=ack,
            retry_count=0,
            max_retries=max_retries,
            created_at=now,
            expires_at=now + timedelta(milliseconds=ttl),
            payload=payload or {},
        )

    # ── Lifecycle ─────────────────────────────────────────────

    def is_expired(self, now: Optional[datetime] = None) -> bool:
        return (now or datetime.now(timezone.utc)) >= self.expires_at

    def is_loop(self, server_id: str, route_history: Optional[list[str]] = None) -> bool:
        """Return True if this envelope has already passed through
        ``server_id``. ``route_history`` is the optional ordered list
        of servers visited (callers can maintain it as a span chain).
        Without history, we conservatively flag a loop only when the
        current server matches the ``source_server_id`` and we've
        already taken at least one hop."""
        if route_history is not None and server_id in route_history:
            return True
        return server_id == self.source_server_id and self.hop_index > 0

    def step(self, next_server_id: str) -> "Envelope":
        """Produce a new envelope advanced one hop. Increments
        ``hop_index``, rotates ``span_id``, sets ``parent_span_id``
        from the previous span, and updates ``current_server_id``.
        Raises ``MaxHopsExceeded`` at the cap."""
        if self.hop_index + 1 >= self.max_hops:
            raise MaxHopsExceeded(
                f"event {self.event_id} reached max_hops={self.max_hops}"
            )
        new = self.model_copy()
        new.parent_span_id = self.span_id
        new.span_id = _gen_id("span")
        new.hop_index = self.hop_index + 1
        new.current_server_id = next_server_id
        new.next_server_id = None
        return new

    def with_retry(self) -> "Envelope":
        """Produce a new envelope with retry_count incremented. Caller
        is responsible for checking ``retry_count <= max_retries``
        before scheduling the retry."""
        new = self.model_copy()
        new.retry_count = self.retry_count + 1
        new.span_id = _gen_id("span")
        new.parent_span_id = self.span_id
        return new

    def size_bytes(self) -> int:
        """Approximate wire size in bytes (UTF-8 JSON)."""
        return len(self.model_dump_json().encode("utf-8"))


# ── Backward-compat shim ───────────────────────────────────────

def from_legacy(
    *,
    event_type: str,
    priority: Priority,
    source_server_id: str,
    payload: dict,
    source_user_id: Optional[str] = None,
    destination_user_id: Optional[str] = None,
    call_id: Optional[str] = None,
    channel_id: Optional[str] = None,
    idempotency_key: Optional[str] = None,
) -> Envelope:
    """Build an envelope from a legacy event dict. Used by the
    federation/emit shim while we migrate handlers to envelope-native
    APIs."""
    # Strip legacy-only keys that don't fit our schema. Whatever's left
    # in payload is preserved.
    payload = dict(payload)  # don't mutate caller's dict
    return Envelope.new(
        event_type=event_type,
        priority=priority,
        source_server_id=source_server_id,
        source_user_id=source_user_id,
        destination_user_id=destination_user_id,
        idempotency_key=idempotency_key,
        call_id=call_id,
        channel_id=channel_id,
        payload=payload,
    )
